core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -492,20 +494,26 @@ private[spark] class BlockManager(
       registerWithExternalShuffleServer()
     }
 
-    hostLocalDirManager =
-      if (conf.get(config.SHUFFLE_HOST_LOCAL_DISK_READING_ENABLED) &&
-          !conf.get(config.SHUFFLE_USE_OLD_FETCH_PROTOCOL)) {
-        externalBlockStoreClient.map { blockStoreClient =>
-          new HostLocalDirManager(
-            futureExecutionContext,
-            conf.get(config.STORAGE_LOCAL_DISK_BY_EXECUTORS_CACHE_SIZE),
-            blockStoreClient,
-            blockManagerId.host,
-            externalShuffleServicePort)
-        }
+    hostLocalDirManager = {
+      val canUseHostLocalReading = conf.get(config.SHUFFLE_HOST_LOCAL_DISK_READING_ENABLED) &&
+        !conf.get(config.SHUFFLE_USE_OLD_FETCH_PROTOCOL)
+      val externalShuffleServiceEnabled = externalBlockStoreClient.isDefined
+      val dynamicAllocationDisabled = !conf.get(config.DYN_ALLOCATION_ENABLED)
+      val dynamicAllocationEnabledWithShuffleTacking = conf.get(config.DYN_ALLOCATION_ENABLED) &&
+        conf.get(config.DYN_ALLOCATION_SHUFFLE_TRACKING_ENABLED)
```

Comment:
As said above, looks like users must be in any of the 3 conditions, i.e. (externalShuffleServiceEnabled || dynamicAllocationDisabled || dynamicAllocationEnabledWithShuffleTacking) is always be true.

Reply:
Thanks for catching it!